### ISSUE TITLE
Uncaught TypeError: Cannot read property 'min' of undefined(…)

### ISSUE THREAD

**Issue by FredCHHsu** [andredumas/techan.js#179]
I am using webpack for bundling JavaScript.
When I tried to import techan, I got this...

Uncaught TypeError: Cannot read property 'min' of undefined(…)   techan.js:1424

I traced the code and found that was from "d3.min",
I tried to directly import d3 and add techan.js to my project.
Still keep receiving error. Any help?

**Comment by briancappello** [comment]
Are you using babel? 

You could try 
```javascript
import * as d3 from 'd3';
```
 but I wasn't able to get that working with events (even though the end of this thread suggests otherwise: https://github.com/d3/d3/issues/2733). 

In the end, the only way I was able to get it working was to include d3 through a classic script tag, outside the flow of webpack. But perhaps you'll have better luck than I did?

**Comment by FredCHHsu** [comment]
@briancappello 
Yes, I am using
`import * as d3 from 'd3';`
as you mentioned.
Still can't make it work within webpack flow.

**Comment by pixelami** [comment]
I believe the problem is here https://github.com/andredumas/techan.js/blob/master/src/techan.js#L6
If `window` exists it insists on referencing d3 as a global.

**Comment by adamrabie** [comment]
@briancappello 

I'm having same issue here but was able to attach it to the window with `new webpack.ProvidePlugin({ "window.d3": "d3" }),` in webpack conf

I still get the error `Uncaught TypeError: Cannot read property 'sourceEvent' of null`

Which is driving me a bit crazy. I've tried using modules individually instead of importing all of d3 and that didn't make a difference. Unfortunate because techan is only package i'm having this babel trouble with. I think its because of two includes of d3 one attached to the window but I cant remove it without techan thinking it doesn't exist.

**Comment by a-hart** [comment]
I am using [https://github.com/SimulatedGREG/electron-vue](url) which is just an electron vue boiler plate, however I was able to get it working with electron/webpack by adding:
 ```
....  extra code removed
plugins: [
new webpack.ProvidePlugin({
       d3: 'd3'
     })]
```

into webpack.renderer.config.js (webpack.config.js is where it normally would be I believe) and then in my index.html where the webpack bundle gets injected I put the following <script> tag: 

  ```
<script>
  var d3 = require('d3')
  </script>
  </head>
  <body>
    <div id="app"></div>
    <!-- webpack builds are automatically injected -->
  </body>
```
This should put d3 on the window.global object. Check the console with 'window.d3' and you should see the d3 object returned and be good to go..

**Comment by albertosantini** [comment]
With `webpack 2.2.1`, `d3 4.5.0` and `techan 0.8.0`, I think this issue (`Uncaught TypeError: Cannot read property 'min' of undefined`) is resolved, following the suggestions in the previous comments.

Adding in the code 

```
import * as d3 from "d3";
...
```

and in `webpack.config.js`

```
...
plugins: [
        new webpack.ProvidePlugin({
            // d3: "d3", // if you add this line you can remove the import in the code
            "window.d3": "d3" // this adds d3 in the window object for techan
        })
```

A different issue is about `Uncaught TypeError: Cannot read property 'sourceEvent' of null`, for instance, you get it, if you use the crosshair feature. This is due to the live bindings of [d3.event](https://github.com/d3/d3-selection#event):

> If you use Babel, Webpack, or another ES6-to-ES5 bundler, be aware that the value of d3.event changes during an event! An import of d3.event must be a live binding, so you may need to configure the bundler to import from D3’s ES6 modules rather than from the generated UMD bundle; not all bundlers observe jsnext:main. Also beware of conflicts with the window.event global.

From a perspective of techan, d3 is loaded in the window object and used as ES5 code, but d3 is loaded, via webpack, as ES2015 (ES6) module: so the d3.event is always null in techan, because it is not binded, via import, to the event of `d3-selection`.

See https://github.com/d3/d3/issues/2733#issuecomment-183685535 and the more recent http://stackoverflow.com/questions/40012016/importing-d3-event-into-a-custom-build-using-rollup.

@andredumas Does it make sense?

**Comment by kalaytan** [comment]
I have the same issue when `import 'techan' `in aurelia.
anybody knows how to fix it?

**Comment by schabluk** [comment]
Same here with

**Comment by gabidi** [comment]
If facing the events issue, then try importing the events module independently 
```js
import * as d3 from 'd3';
import {event as currentEvent} from 'd3';

// if the above doesn't work TRY 
import * as d3 from 'd3';
import {event as currentEvent} from 'd3-selection';
```
Suggestion from:
https://github.com/d3/d3/issues/2733

**Comment by ralyodio** [comment]
`Uncaught Error: Cannot find module 'd3'` from within electron-vue app

edit: Fixed with `import * as d3 from 'd3'` and 
```
  plugins: [
    new webpack.ProvidePlugin({
      'window.d3': 'd3'
    }),
``
inside `.electron-vue/webpack.renderer.config.js`

**Comment by ralyodio** [comment]
`sourceEvent.js? Uncaught TypeError: Cannot read property 'sourceEvent' of null`

i got the chart to load but using the crosshairs causes this problem now (inside vue.js component).

I do not see `sourceEvent` used in the code anywhere.

**Comment by ralyodio** [comment]
Here is an example techan-vue project to debug the issue. I am still getting this error.

https://github.com/chovy/techan-vue

```
# install dependencies
npm install

# serve with hot reload at localhost:8080
npm run dev
```

**Comment by aelkman** [comment]
So no one ever found a solution to this? :/

**Comment by jashmenn** [comment]
@aelkman What fixed it for me was in my webpack file adding:

```
const d3File = path.join(
  __dirname,
  'node_modules/d3/index.js'
);

//....
export default {
  module: {
    rules: [
      // ... other rules
      {
        test: d3File,
        use: [
          {
            loader: 'expose-loader',
            options: 'd3'
          }
        ]
      }
    ]
  },

```

**Comment by aelkman** [comment]
@jashmenn did you put this in the main application component? I have several components files in Vue... it doesn't work for me putting it in the chart component handling the invocation of the chart.

**Comment by jashmenn** [comment]
The basic idea is that one needs `d3` on `window` or `global`, with some setups.

**Comment by slavikdenis** [comment]
Yep, just before you render your app, put this in your index file:
```
import * as d3 from 'd3';
window.d3 = d3;
```

**Comment by AlexanderSergan** [comment]
#### Note for Angular7 users:
You can assign `d3` to `window` object in `src/polyfills.ts` file: 
```typescript
import * as d3 from 'd3'
(window as any).d3 = d3
```

**Comment by simplecodeexamples** [comment]
@AlexanderSergan After trying every possible solution your 

> #### Note for Angular7 users:
> You can assign `d3` to `window` object in `src/polyfills.ts` file:
> 
> ```ts
> import * as d3 from 'd3'
> (window as any).d3 = d3
> ```

After Trying every possible solution this one finally worked for me. @AlexanderSergan Thanks a lot for saving my day.

**Comment by hassanfarahani** [comment]
Hello everybody,
I am new to the web development, vue & d3.. I am working on a project (vuejs & d3). I have the same problem (d3.event is null). would you please tell me in details how I can solve that? I red all of the above comments & solutions, but could not understand what I need to do. Please help me.
Thank you

**Comment by ivanmatiaspascual** [comment]
Here Techan.js wrapper to solve the dependency to window.d3 as a typescript library

https://github.com/ivanmatiaspascual/techan.ts/tree/develop

**Comment by shaf3i** [comment]
if your using **next js**
you can solve it by install **webpack**
and edit **next.config.mjs** file to be : 
```
/** @type {import('next').NextConfig} */
import webpack from "webpack";
const nextConfig = {
  reactStrictMode: true,
  webpack: (config, { isServer }) => {
    if (!isServer) {
      config.plugins.push(
        new webpack.ProvidePlugin({
          "window.d3": "d3",
        })
      );
    }
    return config;
  },
};

export default nextConfig;
```
